python/src/opendp/extras/mbi/_table/__init__.py

```diff
@@ -0,0 +1,548 @@
+    as well as the noise scale and optional threshold for one-way marginals.
+
+    :param input_domain: domain of input data
+    :param input_metric: how to compute distance between datasets
+    :param output_measure: how to measure privacy of release
+    :param d_in: bounds on distances between adjacent input datasets
+    :param d_out: upper bound on the privacy loss
+    :param keys: dictionary of column names and unique categories
+    :param cuts: dictionary of column names and left-open bin edges
+    :param table: ContingencyTable from prior release
+    :param algorithm: configuration for internal estimation algorithm
+    """
+    import_optional_dependency("mbi")
+    import polars as pl
+    import mbi
+
+    plan = _lazyframe_from_domain(input_domain)
+    schema = plan.collect_schema()
+
+    keys_pl: dict[str, pl.Series] = {
+        col: _unique(_with_null(pl.Series(col, keyset, dtype=schema.get(col))), "keys")
+        for col, keyset in (keys or dict()).items()
+    }
+
+    cuts_pl: dict[str, pl.Series] = {
+        col: _unique(_increasing(pl.Series(col, cutset, dtype=schema.get(col))), "cuts")
+        for col, cutset in (cuts or dict()).items()
+    }
+
+    # add cut bin labels to keys
+    def get_categories(cutset):
+        base = cutset.clone().extend(pl.Series([cutset[-1] + 1]))
+        return base.cut(cutset).cat.get_categories().cast(pl.Categorical)  # type: ignore[arg-type]
+
+    keys_pl |= {col: get_categories(cutset) for col, cutset in cuts_pl.items()}
+
+    if table:
+        keys_pl |= table.keys
+        cuts_pl |= table.cuts
+        model = cast(mbi.MarkovRandomField, table.model)
+        marginals = table.marginals.copy()
+        thresholds = table.thresholds.copy()
+    else:
+        model = None
+        marginals = {}
+        thresholds = {}
+
+    if cuts_pl:
+        plan = plan.with_columns(pl.col(c).cut(cutset) for c, cutset in cuts_pl.items() if c in schema)  # type: ignore[arg-type]
+
+    if (QO := RuntimeType.infer(d_out)) != output_measure.distance_type:
+        raise ValueError(f"d_out type ({QO}) must be {output_measure.distance_type}")
+
+    delta = None
+    if isinstance(d_out, tuple):
+        d_out, delta = d_out
+
+    all_keys_known = set(input_domain.columns) <= set(keys_pl)
+    if delta and all_keys_known:
+        message = "delta must be zero because keys and cuts span all columns"
+        raise ValueError(message)
+    if not delta and not all_keys_known:
+        message = "delta must be nonzero because keys and cuts don't span all columns"
+        raise ValueError(message)
+
+    if algorithm.oneway == "unkeyed" and all_keys_known:
+        d_multiway: Union[float | tuple[float, float]] = d_out
+        d_mids = [d_multiway if delta is None else (d_multiway, 0.0)]
+        m_oneway = None
+        scale, threshold = None, None
+
+    else:
+        if algorithm.oneway_split is None:
+            oneway_split = 0.5
+            if algorithm.oneway == "unkeyed":
```

Comment:
Usual worry about hard coded strings. Add a global? Enum with "unkeyed" and "all" might be overkill